src/web/js/trove/d3-lib.js

```diff
@@ -397,70 +410,137 @@
-      image.onload = function () {
-          var opts = {
-              canvas: canvas.node(),
-              image: image
-          };
-          var a = document.createElement('a');
-          a.download = 'sample.png';
-          a.href = drawImage(opts).toDataURL('image/png');
-          a.click();
-
-          // the image's size was doubled everytime we click
-          // the button, so remove all data to prevent this
-          // to happen
-          svgData.remove();
-          canvas.remove();
+  function onSave(detached) {
+    var svgData = detached.append('div').style('display', 'none');
+    var imgsrc = getImageAsURL(detached);
+    var svg = detached.select('svg');
+    var canvas = detached.append('canvas')
+      .style('display', 'none')
+      .attr('width', svg.attr('width'))
+      .attr('height', svg.attr('height'));
+    svgData.html('<img src="' + imgsrc + '">');
+
+    var image = new Image;
+    image.src = imgsrc;
+    image.onload = function () {
+      var opts = {
+        canvas: canvas.node(),
+        image: image
       };
+      var a = document.createElement('a');
+      a.download = 'sample.png';
+      a.href = drawImage(opts).toDataURL('image/png');
+      a.click();
+
+      // somehow the image's size was doubled everytime we click
+      // the button, so remove all data to prevent this
+      // to happen
+      svgData.remove();
+      canvas.remove();
     };
   }
 
-  function callBigBang(detached, dimension, retVal, extra) {
+  function callBigBang(detached, restarter, resizer, windowOptions, dimension, retValFunc, extra) {
+    detached.select('.maing')
+      .append('text')
+      .attr('x', (dimension.marginLeft + dimension.width + dimension.marginRight) / 2)
+      .attr('y', dimension.height + dimension.marginTop + (8 * dimension.marginBottom / 11))
+      .html(libJS.htmlspecialchars(RUNTIME.getField(windowOptions, 'title')))
+      .style({
+        position: 'absolute',
+        'font-size': '8pt',
+        'text-anchor': 'middle'
+      });
+
     detached.selectAll('.overlay').style({
       fill: 'none',
       'pointer-events': 'all',
     });
 
-    RUNTIME.pauseStack(function(restarter) {
-      extra(restarter);
+    if (retValFunc === null) {
+      retValFunc = function(restarter) {
+        imageReturn(detached, restarter, function(x) { return x; });
+      }
+    }
+
+    if (RUNTIME.isPyretFalse(RUNTIME.getField(windowOptions, 'interact'))) {
+      RUNTIME.pauseStack(retValFunc);
+    }
+
+
+    var xscaler = libNum.scaler(
+      dimension.minWindowWidth,
+      dimension.maxWindowWidth,
+      0,
+      1
+    );
+
+    var yscaler = libNum.scaler(
+      dimension.minWindowHeight,
+      dimension.maxWindowHeight,
+      0,
+      1
+    );
+
+    var pauseStack;
+
+    if (RUNTIME.isNothing(restarter)) {
+      pauseStack = RUNTIME.pauseStack;
+    } else {
+      pauseStack = function(cb) { cb(restarter) };
+    }
+
+    pauseStack(function(restarter) {
```

Comment:
I want to write the following function:

```
fun plot(window_size) {
  response-after-user-closes-window = open-plot-window()
  case (response-after-user-closes-window) {
    | user-says-exit => return image
    | user-says-resize => return plot(new-window-size)
  }
}
```

Things are complicated because 1) both opening the plot window and returning an image must be done asynchronously. Thus, `pauseStack` is essential. The simplified code is the following.

```
fun plot(window-size) {
  rt.pauseStack(fun(restarter) {
    open-plot-window(fun(response-after-user-closes-window) {
      case (response-after-user-closes-window) {
        | user-says-exit => 
          var image = new Image;
          image.onload = fu(){
            restarter.resume(image);
          }
        | user-says-resize => plot(new-window-size)
      }
    })
  });
}
```

The problem that I am having is, when I recursively call `plot`, it will hit `pauseStack` again. As you said, after I call `pauseStack`, I am no longer in Pyret stack, so calling it again will result in an error. What I did to avoid this is to write

```
fun plot(window-size, maybe-restarter) {
  my-pause-stack = case(maybe-restarter) {
    | none => rt.pauseStack
    | some(restarter) => fun(callback) { callback(restarter) }
  }
  my-pause-stack(fun(restarter) {
    open-plot-window(fun(response-after-user-closes-window) {
      case (response-after-user-closes-window) {
        | user-says-exit => 
          var image = new Image;
          image.onload = fu(){
            restarter.resume(image);
          }
        | user-says-resize => plot(new-window-size, some(restarter))
      }
    })
  });
}
```

But it's super hacky and might not be the right thing to do. And that's basically the question I'm asking. What's the right thing to do here?